New Module: ec2_vpc_vpn_facts

Madhura-CSI opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
New Module for Facts collection of VPN Connections
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
ec2_vpc_vpn_facts

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.4.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/centos-dev/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```

##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
We often have requirements to gather information about existing VPN Connections in AWS.
Hence, I would like to contribute this module to community.

Sample module call is as follows:
<!--- Paste verbatim command output below, e.g. before and after your change -->
```
- ec2_vpc_vpn_facts:
     filters:
       customer-gateway-id: cgw-e5a52cd5
  register: vpn_connection_info
```

I have written a couple of integration(sanity) tests around this.
Command used to verify the output of tests is:
```
[centos@localhost ansible]$ test/runner/ansible-test integration ec2_vpc_vpn_facts --docker centos7
```
Output file: 
[ec2_vpc_vpn_facts-integration-tests.txt](https://github.com/ansible/ansible/files/1712274/ec2_vpc_vpn_facts-integration-tests.txt)
Please review this PR.

Madhura-CSI committed:
New module: ec2_vpc_vpn_facts

JohnVonNeumann commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn_facts.py:
```diff
@@ -0,0 +1,151 @@
+  register: vpn_conn_facts
+'''
+
+RETURN = '''
```
Forgive me if I'm wrong but I would assume return values should stay consistently laid out like in other facts modules such as [ec2_asg_facts](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ec2_asg_facts.py). I've had a few occasions recently where I've found lacking documentation so I reckon it should be be laid out the same way.

Madhura-CSI replied:
@JohnVonNeumann , thank you for your suggestion. Yes, I do agree with you to keep return values to be consistent with other modules. Through the latest commit, I have corrected these values.
Only for 'customer_gateway_configuration' field, haven't attached sample as it is a big XML string.
Please review and let me know if you would like to see any more changes.

Madhura-CSI committed:
Formatted return values

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn_facts.py:
```diff
@@ -0,0 +1,211 @@
+
+    try:
+        result = json.loads(json.dumps(connection.describe_vpn_connections(**params), default=date_handler))
+    except Exception as e:
```
This should handle specific exceptions that might occur rather than `except Exception`. BotoCoreError, ClientError, maybe ValueError for json.loads/dumps if you think that could fail.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn_facts.py:
```diff
@@ -0,0 +1,211 @@
+
+    region, ec2_url, aws_connect_params = get_aws_connection_info(module, boto3=True)
+
+    connection = boto3_conn(module, conn_type='client', resource='ec2', region=region, endpoint=ec2_url, **aws_connect_params)
```
Since you're using AnsibleAWSModule you can remove the line using get_aws_connection info and this line (and their imports) and replace it with `connection = module.client('ec2')`

s-hertel commented on test/integration/targets/ec2_vpc_vpn_facts/aliases:
```diff
@@ -0,0 +1 @@
+cloud/aws
```
posix/ci/cloud/group4/aws should be added so these tests run in CI.

s-hertel commented on test/integration/targets/ec2_vpc_vpn_facts/tasks/main.yml:
```diff
@@ -0,0 +1,32 @@
+    ec2_vpc_vpn_facts:
+      region: '{{ ec2_region }}'
+      aws_access_key: '{{ aws_access_key }}'
+      aws_secret_key: '{{ aws_secret_key }}'
```
These tests need to take security tokens so they can be run with sts credentials.

willthames replied:
In general, follow the [module guidelines](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#aws-credentials-for-integration-tests) as they cover the detail @s-hertel mentions and more.

s-hertel commented on test/integration/targets/ec2_vpc_vpn_facts/tasks/main.yml:
```diff
@@ -0,0 +1,32 @@
+        - 'result.vpn_connections == [] or result.vpn_connections != []'
+
+  # ============================================================
+  - name: test success with no parameters
```
Could rename this since this test is using a parameter. These tests don't really test the output of the module, just that it can run. Ideally, there would be a VPN created to test vpn_connection_ids. Not a blocker though, more tests than plenty of facts modules have.

Madhura-CSI committed:
integration test with vpn creation and other suggested changes

Madhura-CSI committed:
actual environment creation to test facts module

Madhura-CSI committed:
E324, E327

willthames commented:
Looks like the CI is missing DescribeVPNGateways @s-hertel @mattclay

mnaniwadekar commented:
@willthames , @s-hertel , I have addressed most of the suggestions by latest commits. But, it seems ansibot run failed due to lack of 'DescribeVpnGateways' permissions(as you(@willthames) have described already). 
Please review and let me know if I need to do any change regarding this in IAM policies or if you would like to have any other improvements to be done in module code.

Madhura-CSI commented:
@willthames , @s-hertel , I have addressed most of the suggestions by latest commits. But, it seems ansibot run failed due to lack of 'DescribeVpnGateways' permissions(as you(@willthames) have described already).
Please review and let me know if I need to do any change regarding this in IAM policies or if you would like to have any other improvements to be done in module code.

willthames commented:
@Madhura-CSI this is good to merge (in my opinion at least) once CI is passing - hopefully @mattclay or @s-hertel will resolve that soon.

s-hertel committed:
Add permissions to run ec2_vpc_vpn_facts tests

s-hertel commented:
I have a PR open to update CI. I'm also going to add another test to this branch to help stabilize the tests a little.

s-hertel committed:
Add retry to VPC removal

s-hertel committed:
Use unique name for VGW

mattclay commented:
CI failure due to traceback at `test/integration/targets/ec2_vpc_vpn_facts/tasks/main.yml:83`:

```
Traceback (most recent call last):
  File "/tmp/ansible_HB7PSq/ansible_module_ec2_customer_gateway.py", line 279, in <module>
    main()
  File "/tmp/ansible_HB7PSq/ansible_module_ec2_customer_gateway.py", line 270, in main
    existing['CustomerGateway']['CustomerGatewayId']
  File "/tmp/ansible_HB7PSq/ansible_module_ec2_customer_gateway.py", line 154, in ensure_cgw_absent
    CustomerGatewayId=gw_id
  File "/usr/local/lib/python2.7/dist-packages/botocore/client.py", line 314, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python2.7/dist-packages/botocore/client.py", line 612, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (IncorrectState) when calling the DeleteCustomerGateway operation: The customer gateway is in use.
```

ryansb commented on test/integration/targets/ec2_vpc_vpn_facts/tasks/main.yml:
```diff
@@ -0,0 +1,112 @@
+      bgp_asn: 12345
+      <<: *aws_connection_info
+
+  - name: delete vpn gateway
```
These should be in an `always` block and have a retry to ensure resources are cleaned up.

s-hertel committed:
Always clean up after tests

ryansb commented on test/integration/targets/ec2_vpc_vpn_facts/tasks/main.yml:
```diff
@@ -0,0 +1,114 @@
+    ec2_vpc_vpn:
+      state: absent
+      vpn_connection_id: '{{ vpn.vpn_connection_id }}'
+      <<: *aws_connection_info
```
These need retries/ignore_errors so one teardown fail doesn't stop teardown entirely.

s-hertel committed:
Add retries

s-hertel committed:
remove whitespace

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_vpn_facts.py:
```diff
@@ -0,0 +1,208 @@
+        description: Any tags assigned to the VPN connection.
+        returned: always
+        type: dict
+        sample: {
```
This sample isn't right anymore. The tags are run through `tag_list_to_ansible_dict` now.

ryansb reviewed (changes requested):
Single docs request. Looks great otherwise.

s-hertel commented:
Thanks @Madhura-CSI!
